Add images collector

seb128 opened the pull request:
Get the age and size of the current daily Desktop iso for the series
where we are building those.

iainlane commented on metrics/collectors/images/images_collector.py:
```diff
@@ -0,0 +1,76 @@
+        data = []
+
+        # There are current daily built for those series
+        for serie in self.active_series:
```
nitpick: the single of series is also series, quality English language :-)

seb128 replied:
Thanks, the code has been replaced in the new version which resolves the issue so closing

iainlane commented on metrics/collectors/images/images_collector.py:
```diff
@@ -0,0 +1,76 @@
+        self.lp = Launchpad.login_anonymously(
+            "metrics",
+            "production",
+            launchpadlib_dir=tempfile.mkdtemp(),
```
just a note, `UbuntuQueueMetrics` has the same problem and this can be fixed later. This temp dir might not be cleaned up. On the actual server we run with `PrivateTmp=yes` but if you're testing locally then directories are leaked.

One thought is to run the metrics under `with closing` and let them define a `close()` function which can do any cleanups, or do the same in `run()`, and call `close()` at the end.

iainlane replied:
https://github.com/ubuntu/ubuntu-release-metrics/issues/8

iainlane commented on metrics/collectors/images/images_collector.py:
```diff
@@ -0,0 +1,76 @@
+import requests, datetime, tempfile
+
+DESKTOP_CURRENT_DAILY = "http://cdimage.ubuntu.com/daily-live/current/%s-desktop-%s.iso"
+DESKTOP_LTS_DAILY = "http://cdimage.ubuntu.com/%s/daily-live/current/%s-desktop-%s.iso"
```
Could we maybe collect for `pending` as well as `current` (having that as a tag maybe)? It could be useful to know what's going on with the pending image for example even if promotion is not happening. Like the age there could be telling you that the image is failing to build.

seb128 replied:
The new revision handle current and pending images

iainlane reviewed (changes requested):
Cheers @seb128. This is a great metric to have.

As discussed a bit on IRC at the end of last week, I think we should aim for collecting this data for *all* flavours. The flavour (and image type?) can be a tag value on the submission. Then we can have a drop down, or separate dashboards, letting you view the ISO status for each flavour, as well as maybe one graph which shows all images in a release. That'd let us see if a common update changed the size or made images stop being built or whatever.

In terms of how to get this data, I just remembered that cdimage is available over rsync and you get the mtime and size that way too. So maybe we could do something like `rsync -r --list-only 'rsync://cdimage.ubuntu.com/cdimage/**/daily*'` and then collect the data for every flavour, every release to account for LTS dailies. There's a ton of rsync options to play with to maybe filter to just what's needed and format the output in the right way.

(There might be some work required to open up the firewall, but that should be doable.)

wdyt?

(also some more specific minor feedback is inline)

seb128 commented:
Thanks, I reworked the code to be based on rsync, some comments

* Rsync
-  I spent several hours trying variations of the rsync call and could come to one unified query, especially to do the 'cdimage/*/daily' in a way which would also cover 'cdimage/daily'. I'm not saying it's not possible but I don't want to spend more time on that at the moment, if someone better than me at this want to suggest an improvement that would be welcome though

- I didn't use --list-only but that seems to allow a --out-format and it was making parsing more difficult

- errors could be handled in a better way, it should maybe fail the collect if the commands are not working?

* The update doesn't fix the launchpad closing yet

* There is a bit of guess work on the naming schemes and path but the result seems correct. Unsure of the layout is properly documented?

* I filtered ubuntu-core-16 images which haven't been updated since 2016 and have a weird naming scheme compared to others

* I added a daily_or_current but we could include the information in the flavor as an alternative

Thanks in advance for the review!

iainlane reviewed (changes requested):
Thanks @seb128 I love it. Got some changes to request but I think they're minor.

(Also I rebased/pushed a change, please re-pull before fixing these - and if you want you can use `git rebase -i` to `fixup` my change into yours, it doesn't need to be a separate commit.)

I checked on the server and rsync already seems to be allowed by the firewall!

Will merge this even when I'm off when I see the changes come in. :slightly_smiling_face:

seb128 commented on metrics/collectors/images/images_collector.py:
```diff
@@ -0,0 +1,109 @@
+        rsync_cmd_output = self.rsync_lists_images()
+
+        for rsyncline in rsync_cmd_output.splitlines():
+            if any(ext in rsyncline for ext in IMAGE_FORMATS):
```
Thanks, changed now

seb128 commented on metrics/collectors/images/images_collector.py:
```diff
@@ -0,0 +1,109 @@
+                    if "current" in rsyncline:
+                        daily_or_current = "current"
+                    else:
+                        daily_or_current = "daily"
```
Ups, sorry I initially wanted to call it daily_type but I was unsure it was clear enough so renamed but didn't pay attention while doing so apparently... fixed now to be current_or_pending

seb128 commented on metrics/collectors/images/images_collector.py:
```diff
@@ -0,0 +1,109 @@
+
+                    data.append(
+                        {
+                            "measurement": "daily_images_details",
```
thanks

seb128 commented on metrics/collectors/images/images_collector.py:
```diff
@@ -0,0 +1,109 @@
+        self.active_series = {s.name: s for s in self.ubuntu.series if s.active}
+        self.date_now = datetime.datetime.now()
+
+    def rsync_lists_images(self):
```
thanks

seb128 commented:
Thanks for the review, I consolidated the changes following the review in one commit and force pushed now, hopefully it's right

iainlane commented on metrics/collectors/images/images_collector.py:
```diff
@@ -0,0 +1,111 @@
+                if "current" in rsyncline:
+                    current_or_pending = "current"
+                else:
+                    current_or_pending = "daily"
```
I think I want this string to say `pending` instead of `daily` as well as the variable :upside_down_face:.

seb128 replied:
Indeed, it's another error from my part, I renamed the variable and didn't pay attention, thanks for catching it and fixing!

iainlane commented on metrics/collectors/images/images_collector.py:
```diff
@@ -0,0 +1,111 @@
+        for url in RSYNC_SERVER_REQUESTS:
+            for img in IMAGE_FORMATS:
+                try:
+                    self.log.debug("Rsync listing %s" % url + img)
```
nitpick:

```suggestion
                    self.log.debug("Rsync listing %s%s", url, img)
```

seb128 replied:
Alright, to me they seem equivalent options but if you are a preference let's go with that!

iainlane reviewed (approved):
Cheers. I pushed two more nitpicks. If you like them then you can merge this and the metrics should start being generated automatically (but give it 45 minutes to 1 hour).

seb128 commented:
Great, thanks again for the review, I got the invitation and it's merged now!
